### ISSUE TITLE
Combine several execution in one launch [aka Re-run]

### ISSUE THREAD

**Issue by scabrerao** [reportportal/agent-net-specflow#35]
When a project contains several assemblies and need to be run in parallel but wants to have all results in the same launch, its not possible obtain the launchId from the client/server specflow agent and need to be managed by reportPortal API directly to start new launch before build and at the end close that launch.

![image](https://user-images.githubusercontent.com/3327293/50953631-4d983400-14b4-11e9-82dc-b618cfe528d3.png)

NUnit 3.0.1 
SpecFlow  2.4.0
ReportPortal.SpecFlow 1.4.0

**Comment by nvborisenko** [comment]
Since this is popular question how to use combined launch for several test executions, and given that current way requires deep knowledge of internal `LaunchReporter` implementation, we can resolve it as built-in functionality.

User sets `Launch_Id` as environment variable (or json property), and agent uses it as `LaunchReporter`.

Waiting https://github.com/reportportal/commons-net/pull/14 and https://github.com/reportportal/commons-net/issues/15

**Comment by nvborisenko** [comment]
Done in 1.5.0. @scabrerao please follow readme file.

**Comment by scabrerao** [comment]
Hi!  Thanks for this quick response.

i try it with 1.5.0 but works equal than before... create a new launch per assamblie 
I remove the oldest lines (screenshot in y previous comment) and reading this explanation:

`Combine several execution in one launch
How it can be done:

CI server creates a RP launch and set REPORTPORTAL_LAUNCH_ID environment variable.
Execute tests as usual.
When all tests run, CI server finishes the RP launch.`

i think we're on the same point with the difference that now LaunchId need to be stored in an environment variable to be used by reportPortal.specflow code. 

But my issue, is...  How i could create the LaunchId?,... Is the unique way send the post request to create it and then put request to close it?

I'm trying to have all assemblies results in one lunch even if the execution comes from local computer and not only from CI server.

**Comment by nvborisenko** [comment]
Please share your command how you execute several test assemblies.

**Comment by scabrerao** [comment]
`.\packages\NUnit.ConsoleRunner\tools\nunit3-console.exe .\app1\bin\%ENVIRONMENT%\Tests.dll
.\app2\bin\%ENVIRONMENT%\Tests.dll
.\app3\bin\%ENVIRONMENT%\Tests.dll
.\app4\bin\%ENVIRONMENT%\Tests.dll
 --agents=4`

**Comment by nvborisenko** [comment]
So many thoughts here how to do it, there are good and not good options. The issue is that 4 agents live in 4 separate machine processes. And the question is "who should create a launch?". It's possible to make 4 agents as friends, but we are talking about 1 machine. What if we want to make 4 agents on different machines as friends... In this case we need super-puper global context. Hm. RP server plays this role. @DzmitryHumianiuk may be it will be useful if api will not create a new launch if the same launch is already in progress. In this case all agents will report results into the same launch.

@scabrerao I can propose several the easiest options:
1. Try to use agent for nunit, not for specflow. Because of the execution is managed by nunit, you will see single launch for all test assemblies.
2. Or just add simple console application into your solution, reference to `ReportPortal.Client` package, and implement the logic for starting and stopping launch by yourself. Examples of C# code: https://github.com/reportportal/client-net/blob/master/ReportPortal.Client.Tests/LaunchItem/LaunchItemFixture.cs

@aliaksandrbasau Alex, does it make sense to implement this as built-in feature on agent side? As to me it might be useful only via 3rd party tool.

**Comment by aliaksandrbasau** [comment]
@nvborisenko My first thought was not to try to implement it on the agent side since there is no easy way to know which process is the first (and has to start a launch) and the last (and has to finish the launch). But after giving it a thought it seems possible to do.

When running in multiple-processes mode, each process will try to create a lock file. One that succeeds will create a launch and create a file with launch id. Other processes will wait until the file with launch id is created. Also all processes will write their PIDs (or some random guids) into another file when they start. They will also remove their PIDs from the file when they finish. One that removes the PID last will finish the launch. So the file system will be a synchronization context. Usually I don't like to use file system for such purposes but it should be fine since we are already using it for RP errors.

Your point about the popularity of the feature made me re-think my attitude towards implementing it on the agent side. But still, it's only going to work when tests are executed on a single machine. The solution will not work when tests are executed on multiple machines which is a natural progression (single thread on a single machine -> multiple threads/processes on a single machine -> single/multiple threads/processes on multiple machines). So I still think that it should be implemented as a 3rd party tool but I less sure about it.

**Comment by scabrerao** [comment]
thanks for all comments! 

@nvborisenko i tried your first option (NUnit agent) and for one a **launch everything in one server works fine!** 

I will wait to version 5 as @DzmitryHumianiuk said to implement in multiple machines :)

**Comment by aliaksandrbasau** [comment]
@scabrerao I wouldn't put too much hope into the new re-run functionality (which is great and long-waited) since the issue with synchronization between processes will not be addressed by it. You will still have to know which process should start a launch. Although any process can close the launch as long as all other continue to report test items as re-runs.
But I can imagine at least 5 cases when it can go wrong :smile:
My point is that re-run functionality works best for re-runs but not for reporting test results from multiple processes.

@DzmitryHumianiuk I hope that it will also support reporting under a particular launch (e.g. when `launchId` is specified):
> it will take latest launch with mentioned name, and just report into it

I really like this part:
> no matter is it IN_PROGRESS or FINISHED

**Comment by DzmitryHumianiuk** [comment]
@aliaksandrbasau 
>I hope that it will also support reporting under a particular launch (e.g. when launchId is specified)
yes it will work with certain `launchID` as well.
That is a plan

**Comment by DzmitryHumianiuk** [comment]
refers to https://github.com/reportportal/reportportal/issues/363

**Comment by damianpereira86** [comment]
> So many thoughts here how to do it, there are good and not good options. The issue is that 4 agents live in 4 separate machine processes. And the question is "who should create a launch?". It's possible to make 4 agents as friends, but we are talking about 1 machine. What if we want to make 4 agents on different machines as friends... In this case we need super-puper global context. Hm. RP server plays this role. @DzmitryHumianiuk may be it will be useful if api will not create a new launch if the same launch is already in progress. In this case all agents will report results into the same launch.
> 
> @scabrerao I can propose several the easiest options:
> 
> 1. Try to use agent for nunit, not for specflow. Because of the execution is managed by nunit, you will see single launch for all test assemblies.
> 2. Or just add simple console application into your solution, reference to `ReportPortal.Client` package, and implement the logic for starting and stopping launch by yourself. Examples of C# code: https://github.com/reportportal/client-net/blob/master/ReportPortal.Client.Tests/LaunchItem/LaunchItemFixture.cs
> 
> @aliaksandrbasau Alex, does it make sense to implement this as built-in feature on agent side? As to me it might be useful only via 3rd party tool.

@nvborisenko I'm running into the opposite scenario of Item 1. I'm using NUnit Agent to run several assemblies, and would like each assembly to have it's own Launch, but they are all included in one Launch. Is there a way to accomplish this, other than running each assembly separately?

**Comment by aliaksandrbasau** [comment]
@damianpereira86 I wouldn't implement it on agent side because it couldn't be fully implemented that way. We would still need orchestration at the higher level - create a launch, save its ID, pass the ID to all agents running on different machines, run the tests, wait for all tests to finish and then close the launch.

If the functionality to start and close a launch is implemented on agent side, we would have to start tests just to create or close a launch without actual execution of tests which don't make much sense - running tests doesn't run tests but create/close launch.

So I would say the tests shouldn't do this orchestration. It's a job for CI.

Maybe it would make sense to implement it as a 3rd party tool (e.g. as a `dotnet` tool). The tool could be used for different languages and unit testing frameworks. Although it could be rather difficult to support different agents if we would want the tool to be responsible for passing launch ID to agents as not all agents support specifying launch ID via configuration and they may have different configuration file formats.

**Comment by nvborisenko** [comment]
> I'm using NUnit Agent to run several assemblies, and would like each assembly to have it's own Launch, but they are all included in one Launch.
@damianpereira86 please create a separate issue in `agent-net-nunit` repository. This is topic for discussion, somebody wants to have one single launch for test assemblies, others - different launches. I know that nunit creates single xml report, correct me if I'm wrong.

**Comment by nvborisenko** [comment]
RP v5 is coming with re-run functionality. As expected, user should say before tests execution, that launch should be "appended" into existing one. This is responsibility of user to specify whether new launch should be created, or should not be. So re-run functionality doesn't help us much.

Continuing discussion:
- Agent tries to create global sync context (machine scoped) and all instances are reporting into single global launch
- User creates launch himself and shares launch ID globally (and then finishes launch)

Opinions?

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko for `client-java` we implemented process of self-merging multiple threads/launches.

https://github.com/reportportal/client-java/wiki/How-to-aggregate-Multi-thread-process-execution-in-one-launch

The idea is to create a **local locked file**, which shares active launch ID.
File remains locked while process, which created it, is alive.
Any thread at the starting moment tries to delete it, if failed - read ID from it.
if not failed, and file deleted - it creates launch and lock file.


**For physically distributed executions,** where was an idea to use re-run capabilities.
Re-run be pointed to specific launch.
and can be just in `latest` mode, when it will take latest launch with specified name.

Adding extra call-step as `get launch with name=xxx AND status=IN_PROGRESS` can help to manage it for cross VM execution.

Logic:
- Initialize. Call GET `get launch with name=xxx AND status=IN_PROGRESS`
- IF no launch with this name in progress = create new launch.
- IF there is a launch with this name in progress = use UUID of it for reporting


this can be implemented as additional parameter in agent props.

**Comment by nvborisenko** [comment]
Thumbs up for the feature: if test assemblies are running in parallel (in different threads or different processes), they should be appeared in single launch. This feature should be built-in feature in agent. It can cover machine scoped cases, and doesn't cover distributed execution.

**Comment by itsho** [comment]
Until the proposed solution(s) will be done - I'll kindly ask - Is it possible to bring back the option to set the LaunchId - as suggested in the body of the issue and as suggested in https://github.com/reportportal/agent-net-specflow/issues/12 ?

The reason I'm asking is that it worked **just perfect** before - with older versions of _SpecFlow_, but after upgrading to v3.x this is no longer an option since `LaunchId` is no longer available to be set, and I cannot find a replacement for this amazing feature? 😢

Thanks in advance

**Comment by nvborisenko** [comment]
Setting it from hooks
```csharp
[BeforeTestRun(Order = -30000)]
public static void BeforeTestRunPart()
{
  ReportPortalAddin.Initializing += ReportPortalAddin_Initializing;
}

private static void ReportPortalAddin_Initializing(object sender, InitializingEventArgs e)
{
  e.Config.Properties["Launch:Id"] = 123;
}
```

More properties: https://github.com/reportportal/commons-net/blob/develop/docs/Configuration.md#reporting-experience
